include/alpaka/block/shared/st/BlockSharedMemStMember.hpp

```diff
@@ -0,0 +1,148 @@
+#ifndef NDEBUG
+                        BlockSharedMemStMemberImpl(uint8_t* mem, unsigned int capacity) : m_mem(mem), m_capacity(capacity) {}
+#else
+                        BlockSharedMemStMemberImpl(uint8_t* mem, unsigned int) : m_mem(mem) {}
+#endif
+                        //-----------------------------------------------------------------------------
+                        BlockSharedMemStMemberImpl(BlockSharedMemStMemberImpl const &) = delete;
+                        //-----------------------------------------------------------------------------
+                        BlockSharedMemStMemberImpl(BlockSharedMemStMemberImpl &&) = delete;
+                        //-----------------------------------------------------------------------------
+                        auto operator=(BlockSharedMemStMemberImpl const &) -> BlockSharedMemStMemberImpl & = delete;
+                        //-----------------------------------------------------------------------------
+                        auto operator=(BlockSharedMemStMemberImpl &&) -> BlockSharedMemStMemberImpl & = delete;
+                        //-----------------------------------------------------------------------------
+                        /*virtual*/ ~BlockSharedMemStMemberImpl() = default;
+
+                        template <typename T>
+                        void alloc() const
+                        {
+                            uint8_t* buf = &m_mem[m_allocdBytes];
+                            new (buf) T();
+                            m_allocdBytes += alignPitch<T>();
+#if (defined ALPAKA_DEBUG_OFFLOAD_ASSUME_HOST) && (! defined NDEBUG)
+                            ALPAKA_ASSERT(m_allocdBytes < m_capacity);
+#endif
+                        }
+
+#if BOOST_COMP_GNUC
+    #pragma GCC diagnostic push
+    #pragma GCC diagnostic ignored "-Wcast-align" // "cast from 'unsigned char*' to 'unsigned int*' increases required alignment of target type"
+#endif
+                        template <typename T>
+                        T& getLatestVar() const
+                        {
+                           return *reinterpret_cast<T*>(&m_mem[m_allocdBytes-alignPitch<T>()]);
+                        }
+#if BOOST_COMP_GNUC
+    #pragma GCC diagnostic pop
+#endif
+
+                        void free() const
+                        {
+                            m_allocdBytes = 0u;
+                        }
+
+                    private:
+                        mutable unsigned int m_allocdBytes = 0;
+                        mutable uint8_t* m_mem;
+#ifndef NDEBUG
+                        const unsigned int m_capacity;
+#endif
+
+                        template<typename T>
+                        static constexpr unsigned int alignPitch()
+                        {
+                            return (static_cast<unsigned int>(sizeof(T))/TDataAlignBytes
+                                + static_cast<unsigned int>(sizeof(T))%TDataAlignBytes>0)*TDataAlignBytes;
```

Comment:
Why are we computing this as `unsigned int`? Shouldn't this be a `ptrdiff_t` or `size_t`?

Reply:
All offsets in the shared memory buffer are unsigned int because GPUs have 32-bit registers (64-bit calculations require more registers) and, at least for  NVIDIA, GPU shared memory has 32-bit banks. This small-object-allocator method is not the most efficient way of managing shared memory and I want to keep the performance impact as small as possible on GPUs. This is the reason the function has to returns unsigned int, not size_t (which is 64-bit on modern GPUs).

Since this is a constexpr function, this computation could also be done with size_t, I just found it to be more consistent to do the computation in 32-bit instead of doing it in 64-bit and casting afterwards.

Reply:
In https://github.com/alpaka-group/alpaka/pull/1033/commits/fa6b79a26dd249fdf91903f220e8b04b356ca3a5 this function is replaced by one named `allocPitch()` which is not constexpr, so that one should compute in 32-bit.